### PR TITLE
TLS support for grpcs backend

### PR DESCRIPTION
**Pull request by JLXIA**
1,   Add TLS tests for gRPCs backend
2,  Align with ESPv1 flags for gRPC TLS

### CODE REVIEW

File: docker/generic/start_proxy.py [hunk 1]
```diff
@@ -467,6 +471,47 @@ def make_argparser():
         - Debug HTTP response headers
         ''')
 
+    parser.add_argument(
+        '--transcoding_always_print_primitive_fields',
```
Comment (nareddyt) [inline docker/generic/start_proxy.py]: Please fix the diff

Comment (JLXIA) [reply docker/generic/start_proxy.py]: this is intentionally.  They were added in the deprecated flag section.

File: docker/generic/start_proxy.py [hunk 2]
```diff
@@ -169,6 +169,10 @@ def make_argparser():
         authentication for HTTPS backends. Requires the certificate and
         key files "client.crt" and "client.key" within this path.''')
 
+    parser.add_argument('--ssl_client_cert_chain_file', default=None, help='''
+        The file path for gRPC backend SSL client certificate chain.
```
Comment (qiwzhang) [inline docker/generic/start_proxy.py]: why is this flag only used for gRPC?

File: docker/generic/start_proxy.py [hunk 3]
```diff
@@ -605,6 +621,8 @@ def enforce_conflict_args(args):
         return "Flag --ssl_port is going to be deprecated, please use --ssl_server_cert_path only."
     if args.tls_mutual_auth and args.ssl_client_cert_path:
         return "Flag --tls_mutual_auth is going to be deprecated, please use --ssl_client_cert_path only."
+    if args.ssl_client_cert_chain_file and args.enable_grpc_backend_ssl:
+        return "Flag --grpc_backend_ssl_cert_chain_file is going to be deprecated, please use ssl_client_cert_chain_file only."
```
Comment (qiwzhang) [inline docker/generic/start_proxy.py]: return error doesn't match the flag it is checking.

File: docker/generic/start_proxy.py (hunk 2 above)
Comment (JLXIA) [reply docker/generic/start_proxy.py]: You are right, I missed it.

File: docker/generic/start_proxy.py (hunk 3 above)
Comment (JLXIA) [reply docker/generic/start_proxy.py]: updated.

File: docker/generic/start_proxy.py [hunk 4]
```diff
@@ -529,46 +574,17 @@ def make_argparser():
...
-        cannot control the query parameters and do not know them beforehand.
-        Otherwise use ignored_query_parameters. Defaults to false.
-        ''')
+    parser.add_argument('--enable_grpc_backend_ssl',
+        action='store_true', help='''
+        This flag added for backward compatible for ESPv1 and will be deprecated.
+        Enable SSL for gRPC backend. ESPv2 auto enables SSL if schema `grpcs` is
+        detected.''')
 
+    parser.add_argument('--grpc_backend_ssl_root_certs_file',
```
Comment (qiwzhang) [inline docker/generic/start_proxy.py]: are we going to support 

grpc_backend_ssl_private_key_file
grpc_backend_ssl_cert_chain_file

Comment (JLXIA) [reply docker/generic/start_proxy.py]: yes,  will follow up in next PR.

File: docker/generic/start_proxy.py [hunk 5]
```diff
@@ -605,6 +621,8 @@ def enforce_conflict_args(args):
         return "Flag --ssl_port is going to be deprecated, please use --ssl_server_cert_path only."
     if args.tls_mutual_auth and args.ssl_client_cert_path:
         return "Flag --tls_mutual_auth is going to be deprecated, please use --ssl_client_cert_path only."
+    if args.ssl_client_cert_chain_file and args.enable_grpc_backend_ssl and args.grpc_backend_ssl_cert_chain_file:
```
Comment (TAOXUY) [inline docker/generic/start_proxy.py]: Where did you define `args.grpc_backend_ssl_cert_chain_file`?

File: tests/start_proxy/start_proxy_test.py [hunk 6]
```diff
@@ -129,7 +129,7 @@ def test_gen_proxy_config(self):
               '--ssl_server_cert_path', '/etc/nginx/ssl',
               '--listener_port', '443',
               ]),
-            # ssl_backend_path specified
+            # ssl_client_cert_path specified
             (['-R=managed','--listener_port=8080',  '--disable_tracing',
               '--ssl_client_cert_path=/etc/endpoint/ssl'],
```
Comment (TAOXUY) [inline tests/start_proxy/start_proxy_test.py]: Add all the error cases for flag conflicts.

Comment (JLXIA) [reply tests/start_proxy/start_proxy_test.py]: Done

File: docker/generic/start_proxy.py (hunk 5 above)
Comment (JLXIA) [reply docker/generic/start_proxy.py]: you are right, I messed up the flags. 

Changed to "ssl_client_root_certs_file" for new flag name, corresponding legacy "grpc_backend_ssl_root_certs_file".

File: docker/generic/start_proxy.py [hunk 7]
```diff
@@ -605,6 +621,8 @@ def enforce_conflict_args(args):
         return "Flag --ssl_port is going to be deprecated, please use --ssl_server_cert_path only."
     if args.tls_mutual_auth and args.ssl_client_cert_path:
         return "Flag --tls_mutual_auth is going to be deprecated, please use --ssl_client_cert_path only."
+    if args.ssl_client_root_certs_file and args.enable_grpc_backend_ssl and args.grpc_backend_ssl_root_certs_file:
```
Comment (TAOXUY) [inline docker/generic/start_proxy.py]: This is non-empty by default so maybe we don't need put in the condition.

File: tests/integration_test/dynamic_routing_integration_test.go [hunk 8]
```diff
@@ -720,3 +722,73 @@ func TestDynamicBackendRoutingMutualTLS(t *testing.T) {
...
+			method:         "/v1/shelves/200?key=api-key",
+			wantResp:       `{"id":"200","theme":"Classic"}`,
+		},
+		{
+			desc:                "gRPC client calling gRPCs remote backend fail",
+			clientProtocol:      "grpc",
+			method:              "GetShelf",
+			useWrongBackendCert: true,
+			header:              http.Header{"x-api-key": []string{"api-key"}},
+			wantError:           "Unavailable",
```
Comment (nareddyt) [inline tests/integration_test/dynamic_routing_integration_test.go]: Nit: "503 Service Unavailable"

File: tests/integration_test/dynamic_routing_integration_test.go [hunk 9]
```diff
@@ -720,3 +722,73 @@ func TestDynamicBackendRoutingMutualTLS(t *testing.T) {
...
+			desc:           "gRPC client calling gRPCs remote backend succeed",
+			clientProtocol: "grpc",
+			method:         "GetShelf",
+			header:         http.Header{"x-api-key": []string{"api-key"}},
+			wantResp:       `{"id":"100","theme":"Kids"}`,
+		},
+		{
+			desc:           "Http client calling gRPCs remote backend succeed",
+			clientProtocol: "http",
+			method:         "/v1/shelves/200?key=api-key",
```
Comment (nareddyt) [inline tests/integration_test/dynamic_routing_integration_test.go]: Nit: Consider changing variable name or using a different one. `method` makes sense for grpc but not for http

File: tests/integration_test/dynamic_routing_integration_test.go [hunk 10]
```diff
@@ -720,3 +722,73 @@ func TestDynamicBackendRoutingMutualTLS(t *testing.T) {
...
+			wantResp:       `{"id":"200","theme":"Classic"}`,
+		},
+		{
+			desc:                "gRPC client calling gRPCs remote backend fail",
```
Comment (nareddyt) [inline tests/integration_test/dynamic_routing_integration_test.go]: Please clarify in the test description that the failure is due to it having an incorrect cert.

File: docker/generic/start_proxy.py (hunk 7 above)
Comment (JLXIA) [reply docker/generic/start_proxy.py]: Done

File: tests/integration_test/dynamic_routing_integration_test.go (hunk 8 above)
Comment (JLXIA) [reply tests/integration_test/dynamic_routing_integration_test.go]: This is gRPC, no HTTP status code returned.

File: tests/integration_test/dynamic_routing_integration_test.go (hunk 9 above)
Comment (JLXIA) [reply tests/integration_test/dynamic_routing_integration_test.go]: Done

File: tests/integration_test/dynamic_routing_integration_test.go (hunk 10 above)
Comment (JLXIA) [reply tests/integration_test/dynamic_routing_integration_test.go]: Done

File: tests/start_proxy/start_proxy_test.py [hunk 11]
```diff
@@ -138,6 +138,24 @@ def test_gen_proxy_config(self):
...
+            (['-R=managed','--listener_port=8080',  '--disable_tracing',
+              '--ssl_client_root_certs_file=/etc/endpoints/ssl/ca-certificates.crt' ],
+             ['bin/configmanager', '--logtostderr',
+              '--backend_address', 'http://127.0.0.1:8082',
+              '--rollout_strategy', 'managed', '--v', '0',
+              '--listener_port', '8080', '--root_certs_path',
+              '/etc/endpoints/ssl/ca-certificates.crt', '--disable_tracing'
+              ]), 
+            # legacy enable_grpc_backend_ssl specified
+            (['-R=managed','--listener_port=8080',  '--disable_tracing',
```
Comment (TAOXUY) [inline tests/start_proxy/start_proxy_test.py]: add unit test of non default grpc_backend_ssl_root_certs_file

File: tests/start_proxy/start_proxy_test.py [hunk 12]
```diff
@@ -138,6 +138,24 @@ def test_gen_proxy_config(self):
               '--listener_port', '8080', '--ssl_client_cert_path',
```
Comment (TAOXUY) [inline tests/start_proxy/start_proxy_test.py]: add conflict test of enable_grpc_backend_ssl and ssl_client_root_certs_file

File: tests/start_proxy/start_proxy_test.py (hunk 11 above)
Comment (JLXIA) [reply tests/start_proxy/start_proxy_test.py]: Ack. Not necessary.

File: tests/start_proxy/start_proxy_test.py (hunk 12 above)
Comment (JLXIA) [reply tests/start_proxy/start_proxy_test.py]: already done in line 342

**Pull request merged**